src/Compiler/Microsoft.AspNetCore.Razor.Language/src/MetadataCollection.cs

```diff
@@ -409,24 +387,67 @@ public override bool Equals(MetadataCollection other)
 
         protected override int ComputeHashCode()
         {
-            var hash = HashCodeCombiner.Start();
-
-            hash.Add(_key1, StringComparer.Ordinal);
-            hash.Add(_value1, StringComparer.Ordinal);
+            return _count switch
+            {
+                1 => ComputeHashCodeForOneItem(),
+                2 => ComputeHashCodeForTwoItems(),
+                _ => ComputeHashCodeForThreeItems()
+            };
 
-            if (_count > 1)
+            int ComputeHashCodeForOneItem()
             {
-                hash.Add(_key2, StringComparer.Ordinal);
-                hash.Add(_value2, StringComparer.Ordinal);
+                var hash = HashCodeCombiner.Start();
+
+                hash.Add(_key1, StringComparer.Ordinal);
+                hash.Add(_value1, StringComparer.Ordinal);
+
+                return hash.CombinedHash;
             }
 
-            if (_count > 2)
+            int ComputeHashCodeForTwoItems()
             {
-                hash.Add(_key3, StringComparer.Ordinal);
-                hash.Add(_value3, StringComparer.Ordinal);
+                var hash = HashCodeCombiner.Start();
+
+                if (string.CompareOrdinal(_key1, _key2) < 0)
+                {
+                    hash.Add(_key1, StringComparer.Ordinal);
+                    hash.Add(_value1, StringComparer.Ordinal);
+                    hash.Add(_key2, StringComparer.Ordinal);
+                    hash.Add(_value2, StringComparer.Ordinal);
+                }
+                else
+                {
+                    hash.Add(_key2, StringComparer.Ordinal);
+                    hash.Add(_value2, StringComparer.Ordinal);
+                    hash.Add(_key1, StringComparer.Ordinal);
+                    hash.Add(_value1, StringComparer.Ordinal);
+                }
+
+                return hash.CombinedHash;
             }
 
-            return hash.CombinedHash;
+            int ComputeHashCodeForThreeItems()
+            {
+                var hash = HashCodeCombiner.Start();
+
+                using var _ = ListPool<KeyValuePair<string, string>>.GetPooledObject(out var list);
```

Comment:
We could make this better on core by using a `Span`, `stackalloc`, and `MemoryExtensions.Sort`, but the `Sort` APIs are only available on .NET 5 and newer. Up to you whether you want to go the `#ifdef` route.

Reply:
`Microsoft.AspNetCore.Razor.Language` is not multi-targeted. It only targets `netstandard2.0`, so an `#ifdef` is not possible. Of course, multi-targeting the compiler is out of the scope of this PR. ;-)